labs/a10.md

```diff
@@ -0,0 +1,369 @@
+
+`docker run hello-world`
+
+You should see some friendly output like so:
+
+```
+Unable to find image 'hello-world:latest' locally
+latest: Pulling from library/hello-world
+b04784fba78d: Already exists 
+Digest: sha256:f3b3b28a45160805bb16542c9531888519430e9e6d6ffc09d72261b0d26ff74f
+Status: Downloaded newer image for hello-world:latest
+
+Hello from Docker!
+This message shows that your installation appears to be working correctly.
+
+To generate this message, Docker took the following steps:
+ 1. The Docker client contacted the Docker daemon.
+ 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
+ 3. The Docker daemon created a new container from that image which runs the
+    executable that produces the output you are currently reading.
+ 4. The Docker daemon streamed that output to the Docker client, which sent it
+    to your terminal.
+
+To try something more ambitious, you can run an Ubuntu container with:
+ $ docker run -it ubuntu bash
+
+Share images, automate workflows, and more with a free Docker ID:
+ https://cloud.docker.com/
+
+For more examples and ideas, visit:
+ https://docs.docker.com/engine/userguide/
+ ```
+
+Some [quick definitions from Docker’s 
+website:](https://docs.docker.com/get-started/#a-brief-explanation-of-containers)
+
+An **image** is a lightweight, stand-alone, executable package that includes everything needed to run a piece of 
+software,
+including the code, a runtime, libraries, environment variables, and config files.
+
+A **container** is a runtime instance of an image—what the image becomes in memory when actually executed.
+It runs completely isolated from the host environment by default, only accessing host files and ports if configured to 
+do so.
+
+Be sure to read through the output from running the hello-world image to get an understanding of what the Docker daemon 
+was doing.
+
+### Running an interactive container
+We’re now going to walk you through running a container interactively. This is useful if you ever need to play around 
+and
+install stuff on a bare system without messing up your current system. Try running the following command:
+
+`docker run -it ubuntu:xenial /bin/bash`
+
+The `-i` flag tells docker to keep `STDIN` open to your container, and the `-t` flag allocates a 
+[pseudo-TTY](https://en.wikipedia.org/wiki/Pseudoterminal) for you. Basically you need both for you to have a way to 
+enter text and have this display properly. At the end of the command, `/bin/bash` is just the command you want to run 
+once the container starts up. Try installing some packages from `apt` or just play around. It should look like a bare 
+Linux system.
+
+You can exit the container with `CTRL+D`.
+
+Notice how even though your VM is running Debian, you were able to run the Xenial version of Ubuntu in a container. If 
+you are curious about other variants of Linux, you can run a lot of them inside containers as well! You can try running 
```

Comment:
will add something here explaining that this works because every one of these OSes uses the Linux kernel, and we obviously can't run Windows or macOS in a docker container